File: aemet/SergioDeIscar-aemet-java/SergioDeIscar-aemet-java/src/main/java/org/example/repositories/dupla/DuplaExtension.java
package org.example.repositories.dupla;

import org.example.models.Dupla;
import org.example.repositories.CrudRepository;

import java.time.LocalDate;
import java.util.List;
import java.util.Map;

public interface DuplaExtension extends CrudRepository<Dupla, String> {
    Map<LocalDate, Map<String, Dupla>> maxTemPorLugar();
    Map<LocalDate, Map<String, Dupla>> minTemPorLugar();
    Map<String, Dupla> maxTemPorProvincia();
    Map<String, Dupla> minTemPorProvincia();
    Map<String, Double> mediaPorProvincia();
    Map<LocalDate, Map<String, Double>> mediaPrecipitacionPorDiaYProvincia();
    Double mediaTemMadrid();
    Double mediaTemMax();
    Double mediaTemMin();

    /**
     * Lugares donde la máxima ha sido antes de las 15:00 por día
     */
    Map<LocalDate, List<String>> maxTemAntes();

    /**
     * Lugares donde la mínima ha sido después de las 17:30 por día
     */
    Map<LocalDate, List<String>> minTemDespues();

    Dupla maxTemMadrid(LocalDate day);
    Dupla minTemMadrid(LocalDate day);
    Double mediaPrecipitacionMadrid(LocalDate day);
}

File: aemet/SergioDeIscar-aemet-kotlin/SergioDeIscar-aemet-kotlin/src/main/kotlin/repositories/dupla/DuplaExtension.kt
package repositories.dupla

import models.Dupla
import models.Informe
import repositories.CrudRepository
import java.time.LocalDate

interface DuplaExtension: CrudRepository<Dupla, String> {
    fun maxTemPorLugar(): Map<LocalDate, Map<String, Double>>
    fun minTemPorLugar(): Map<LocalDate, Map<String, Double>>
    fun maxTemPorProvincia(): Map<String, Double>
    fun minTemPorProvincia(): Map<String, Double>
    fun mediaPorProvincia(): Map<String, Double>
    fun mediaPrecipitacionPorDiaYProvincia(): Map<LocalDate, Map<String, Double>>
    fun mediaTemMadrid(): Double
    fun mediaTemMax(): Double
    fun mediaTemMin(): Double

    /**
     * Lugares donde la máxima ha sido antes de las 15:00 por día
     */
    fun maxTemAntes(): Map<LocalDate, List<String>>

    /**
     * Lugares donde la mínima ha sido después de las 17:30 por día
     */
    fun minTemDespues(): Map<LocalDate, List<String>>

    fun maxTemMadrid(day: LocalDate): Dupla
    fun minTemMadrid(day: LocalDate): Dupla
    fun mediaPrecipitacionMadrid(day: LocalDate): Double
}